File: Rust/src/year2019/intcode_computer/computer.rs
use crate::year2019::intcode_computer::instruction::{Instruction, Mode};
use crate::year2019::intcode_computer::opcode::Opcode;
use std::io::{self, Write, Read};

#[allow(dead_code)]
#[derive(PartialEq, Eq, Clone)]
pub enum State
{
    Running,
    WaitingForInput,
    PushedOutput,
    Break,
    Halt,
}

pub struct Computer
{
    memory: Vec<i64>,

    instruction_pointer: u64,
    relative_base_pointer: i64,
    state: State,

    //I/O
    pub input:  Vec<i64>,
    pub output: Vec<i64>,

    //Settings:
    pub print_disassembly: bool,
    pub print_output: bool,

    pub break_pointer: Option<u64>,
}

impl Computer
{
    //Constructors ============================================================================================================================
    fn new() -> Self
    {
        Computer
        {
            memory: Vec::new(),
            instruction_pointer: 0,
            relative_base_pointer: 0,
            state: State::Running,

            input: Vec::new(),
            output: Vec::new(),

            print_disassembly: true,
            print_output: true,
            break_pointer: None,
        }
    }

    pub fn reset_from_str(&mut self, str: &str)
    {
        self.memory.clear();
        for num in str.split(',')
        {
            self.memory.push(num.parse::<i64>().unwrap());
        }
        self.instruction_pointer = 0;
        self.relative_base_pointer = 0;
        self.state = State::Running;

        self.input.clear();
        self.output.clear();
    }

    pub fn from_str(str: &str) -> Self
    {
        let mut computer = Computer::new();
        for num in str.split(',')
        {
            computer.memory.push(num.parse::<i64>().unwrap());
        }
        return computer;
    }
    //============================================================================================================================

    //Helpers ============================================================================================================================
    #[allow(dead_code)]
    pub fn memory_to_string(&self) -> String
    {
        return self.memory.iter().map(|x| x.to_string()).collect::<Vec<String>>().join(",");
    }

    pub fn memory_read(&self, mode: &Mode, value: i64) -> i64
    {
        match mode
        {
            Mode::Immediate => return value,
            Mode::Relative =>
            {
                let address = self.relative_base_pointer + value;
                if address >= self.memory.len() as i64
                {
                    return 0;
                }
                return self.memory[address as usize];
            }
            Mode::Position => {
                if value >= self.memory.len() as i64
                {
                    return 0;
                }
                return self.memory[value as usize]
            },
            //_ => panic!("Memory read for this mode not implemented.")
        }
    }

    pub fn memory_write(&mut self, location: i64, value: i64)
    {
        //If memory is too small, grow it to the requested size
        if location >= self.memory.len() as i64
        {
            self.memory.resize((location + 1) as usize, 0);
        }
        self.memory[location as usize] = value;
    }

    pub fn disassemble_program(&self)
    {
        let mut ptr = 0;
        loop
        {
            let instruction = Instruction::parse(ptr, &self.memory);
            println!("{}", instruction.disassemble(ptr, self.relative_base_pointer, &self));
            ptr += instruction.size as u64;
            if instruction.opcode == Opcode::Halt
            {
                break;
            }
        }
    }

    pub fn get_state(&self) -> State
    {
        return self.state.clone();
    }
    //============================================================================================================================

    //Code execution ============================================================================================================================
    pub fn run(&mut self)
    {
        loop
        {
            let state = self.step();
            match state
            {
                State::Running => {/*just keep running*/}
                State::WaitingForInput =>
                {
                    print!("in: ");
                    io::stdout().flush().unwrap();
                    let mut buffer = String::new();
                    io::stdin().read_line(&mut buffer).unwrap();
                    let num = buffer.trim().parse::<i64>().unwrap();
                    self.input.push(num);
                    println!();
                }
                State::PushedOutput =>
                {
                    if self.print_output
                    {
                        println!("out: {}", self.output[0]);
                        self.output.remove(0);
                    }
                }
                State::Break =>
                {
                    println!("computer entered break mode. Press any key to continue.");
                    io::stdout().flush().unwrap();

                    //Read so a keypress is required
                    let mut buffer = [0, 10];
                    io::stdin().read(&mut buffer).unwrap();

                    //Clear this breakpoint to prevent a loop
                    self.break_pointer = None;
                    self.state = State::Running;
                }
                State::Halt => return,
            }
        }
    }

    pub fn step(&mut self) -> State
    {
        //========================================================================================================================================================================
        //Parsing

        //Clear any flags recieved previously
        self.state = State::Running;

        match self.break_pointer
        {
            Some(number) =>
            {
                 if self.instruction_pointer == number
                 {
                     self.state = State::Break;
                     return self.state.clone();
                 }
            }
            _ => {}
        }


        let instruction = Instruction::parse(self.instruction_pointer, &self.memory);

        if self.print_disassembly
        {
            println!("{}", instruction.disassemble(self.instruction_pointer, self.relative_base_pointer, &self));
        }


        //========================================================================================================================================================================
        //Pre calculation: calculating a couple things here that will make life easier when executing the instructions


        //Instead of fetching the arguments from memory at every twist and turn, we'll fetch them beforehand
        //we will only be fetching the correct amount else we might be reading outside of our memory
        //The variables must always exist
        let mut numbers: Vec<i64> = vec!{0,0,0};
        if instruction.opcode != Opcode::Halt
        {
            for i in 0..(instruction.size-1) as usize
            {
                numbers[i] = self.memory_read(&instruction.argument_modes[i], instruction.arguments[i]);
            }
        }

        //write address - even though early on it is suggested that the write address is always in position mode, it can also be relative.
        //We can figure out the write address beforehand, using the count of arguments.
        let mut write_address = 0;
        if instruction.argument_count > 0
        {
            //On instructions that can write, the last argument is always the write address
            let last_argument_index = (instruction.argument_count-1) as usize;
            write_address = match instruction.argument_modes[last_argument_index]
            {
                Mode::Relative => self.relative_base_pointer + instruction.arguments[last_argument_index],
                _ => instruction.arguments[last_argument_index]
            }
        }


        //Note: when writing to memory, the raw value from the instruction is used.
        //That is because the output location should always be seen as immediate even if it's mode is position.
        //Take this for example: 1,0,0,0,99
        //Will try to read the 3rd 0 and write the result to location 1. The result should be in location 0.

        //========================================================================================================================================================================
        //Executing

        match instruction.opcode
        {
            Opcode::Add =>
            {
                self.memory_write(write_address, numbers[0] + numbers[1]);
            }
            Opcode::Multiply =>
            {
                self.memory_write(write_address, numbers[0] * numbers[1]);
            }
            Opcode::Input =>
            {
                if self.input.len() > 0
                {
                    let input = self.input[0];
                    //Erase the value we just used as input
                    self.input.remove(0);

                    self.memory_write(write_address, input);
                }
                else
                {
                    //We have no input: Do not increment instruction pointer, break out so that input can be supplied
                    self.state = State::WaitingForInput;
                    return self.state.clone();
                }
            }
            Opcode::Output =>
            {
                //Notify consumers of step that we pushed an output, do increment the instruction pointer so that we don't loop.
                self.output.push(numbers[0]);
                self.state = State::PushedOutput;
            }
            Opcode::JumpIfTrue =>
            {
                if numbers[0] != 0
                {
                    self.instruction_pointer = numbers[1] as u64;
                    //Exit without incrementing the instruction pointer
                    return self.state.clone();
                }
            }
            Opcode::JumpIfFalse =>
            {
                if numbers[0] == 0
                {
                    self.instruction_pointer = numbers[1] as u64;
                    //Exit without incrementing the instruction pointer
                    return self.state.clone();
                }
            }
            Opcode::LessThan =>
            {
                let mut num = 0;
                if numbers[0] < numbers[1]
                {
                    num = 1;
                }
                self.memory_write(write_address, num);
            }
            Opcode::Equals =>
            {
                let mut num = 0;
                if numbers[0] == numbers[1]
                {
                    num = 1;
                }
                self.memory_write(write_address, num);
            }
            Opcode::AdjustRelativeBase =>
            {
                self.relative_base_pointer += numbers[0];
            }

            Opcode::Halt =>
            {
                self.state = State::Halt;
            }
            //_ => panic!("Opcode not implemented."),
        }
        self.instruction_pointer += instruction.size as u64;
        return self.state.clone();
    }
    //============================================================================================================================
}

File: C#/src/Years/Year2019/IntCodeComputer/Computer.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

namespace Years.Year2019.IntCodeComputer
{
    public enum State
    {
        Running,
        WaitingForInput,
        PushedOutput,
        Break,
        Halt,
    }

    public class Computer
    {
        public List<long> Memory = new List<long>();

        public long InstructionPointer = 0;
        public long RelativeBasePointer = 0;
        public State State = State.Running;

        //I/O
        public List<long> Input = new List<long>();
        public List<long> Output = new List<long>();

        //Settings:
        public bool PrintDisassembly = true;
        public bool PrintOutput = true;

        public long? BreakPointer = null;


        #region Constructors ========================================================================================================
        public Computer(string program)
        {
            foreach (var s in program.Split(','))
            {
                Memory.Add(long.Parse(s));
            }
        }

        public void Reset(string program)
        {
            Memory.Clear();
            foreach (var s in program.Split(','))
            {
                Memory.Add(long.Parse(s));
            }

            InstructionPointer = 0;
            RelativeBasePointer = 0;
            State = State.Running;

            Input.Clear();
            Output.Clear();
        }
        #endregion


        #region Helpers ========================================================================================================

        public string MemoryToString()
        {
            return string.Join(',', Memory);
        }


        public long ReadMemory(Mode mode, long value)
        {
            switch(mode)
            {
                case Mode.Immediate:
                    return value;
                case Mode.Relative:
                    var address = RelativeBasePointer + value;
                    //Uninitialized memory is treated as 0.
                    if (address >= Memory.Count)
                    {
                        return 0;
                    }
                    return Memory[(int)address];
                case Mode.Position:
                    if (value >= Memory.Count)
                    {
                        return 0;
                    }
                    return Memory[(int)value];
                default: throw new Exception("Reading memory for unsupported mode.");
            }
        }


        public void WriteMemory(long address, long value)
        {
            //If memory is too small, grow it to the requested size
            if (address >= Memory.Count)
            {
                Memory.AddRange(new long[address+1 - Memory.Count]);
            }
            Memory[(int)address] = value;
        }

        public void DisassembleProgram()
        {
            long i = 0;
            while (true)
            {
                var instruction = Instruction.ParseInstruction(i, Memory);
                Console.WriteLine(instruction.Disassemble(this));
                i += (long) instruction.Size;
                if (instruction.Opcode == Opcode.Halt)
                {
                    return;
                }
            }
        }

        #endregion


        #region Code Execution ========================================================================================================
        /// <summary>
        /// Example run function that takes input from the console and pushes output to the console
        /// Consume the Step function to create custom behavior for I/O
        /// </summary>
        public void Run()
        {
            while(true)
            {
                var state = Step();
                switch(State)
                {
                    case State.WaitingForInput:
                        Console.WriteLine("in: ");
                        Console.Out.Flush();
                        var input = Console.ReadLine();
                        Input.Add(long.Parse(input));
                        Console.WriteLine();
                        break;

                    case State.PushedOutput:
                        if (PrintOutput)
                        {
                            Console.WriteLine($"out: {Output[0]}");
                            Output.RemoveAt(0);
                        }
                        break;

                    case State.Break:
                        Console.WriteLine("Breakpoint hit. Press any key to continue.");
                        Console.ReadKey();
                        State = State.Running;
                        break;

                    case State.Halt:
                        return;
                }
            }
        }



        /// <summary>
        /// Executes a single instruction and returns the state after this instruction
        /// </summary>
        /// <returns></returns>
        public State Step()
        {
            //Clear any flags recieved previously
            State = State.Running;

            //========================================================================================================================================================================
            //Parsing

            if (BreakPointer != null)
            {
                if (InstructionPointer == BreakPointer.Value)
                {
                    Debugger.Break();
                }
            }


            var instruction = Instruction.ParseInstruction(InstructionPointer, Memory);
            if (PrintDisassembly)
            {
                Console.WriteLine(instruction.Disassemble(this));
            }


            //========================================================================================================================================================================
            //Pre calculation: calculating a couple things here that will make life easier when executing the instructions
            
            //Instead of fetching the arguments from memory at every twist and turn, we'll fetch them beforehand
            //we will only be fetching the correct amount else we might be reading outside of our memory
            //The variables must always exist
            var numbers = new List<long>() {0, 0, 0};
            if (instruction.Opcode != Opcode.Halt)
            {
                for (int i = 0; i < instruction.Size - 1; i++)
                {
                    numbers[i] = ReadMemory(instruction.ArgumentModes[i], instruction.Arguments[i]);
                }
            }


            //write address - even though early on it is suggested that the write address is always in position mode, it can also be relative.
            //We can figure out the write address beforehand, using the count of arguments.
            long writeAddress = 0;
            if (instruction.ArgumentCount > 0)
            {
                var lastArgumentIndex = instruction.ArgumentCount - 1;
                switch (instruction.ArgumentModes[(int)lastArgumentIndex])
                {
                    case Mode.Relative:
                        writeAddress = RelativeBasePointer + instruction.Arguments[(int)lastArgumentIndex];
                        break;
                    default:
                        writeAddress = instruction.Arguments[(int)lastArgumentIndex];
                        break;
                }
            }



            //Note: when writing to memory, the raw value from the instruction is used.
            //That is because the output location should always be seen as immediate even if it's mode is position.
            //Take this for example: 1,0,0,0,99
            //Will try to read the 3rd 0 and write the result to location 1. The result should be in location 0.

            //========================================================================================================================================================================
            //Executing
            switch (instruction.Opcode)
            {
                case Opcode.Add:
                    WriteMemory(writeAddress, numbers[0] + numbers[1]);
                    break;
                case Opcode.Multiply:
                    WriteMemory(writeAddress, numbers[0] * numbers[1]);
                    break;


                case Opcode.Input:
                    if (Input.Count > 0)
                    {
                        var input = Input[0];
                        //Erase the value we just used as input
                        Input.RemoveAt(0);
                        WriteMemory(writeAddress, input);
                    }
                    else
                    {
                        //We have no input: Do not increment instruction pointer, break out so that input can be supplied
                        State = State.WaitingForInput;
                        return State;
                    }
                    break;
                case Opcode.Output:
                    //Notify consumers of step that we pushed an output, do increment the instruction pointer so that we don't loop.
                    Output.Add(numbers[0]);
                    State = State.PushedOutput;
                    break;


                case Opcode.JumpIfTrue:
                    if (numbers[0] != 0)
                    {
                        InstructionPointer = numbers[1];
                        //Exit without incrementing the instruction pointer
                        return State;
                    }
                    break;
                case Opcode.JumpIfFalse:
                    if (numbers[0] == 0)
                    {
                        InstructionPointer = numbers[1];
                        //Exit without incrementing the instruction pointer
                        return State;
                    }
                    break;


                case Opcode.LessThan:
                    long num = 0;
                    if (numbers[0] < numbers[1])
                    {
                        num = 1;
                    }
                    WriteMemory(writeAddress, num);
                    break;
                case Opcode.Equals:
                    long temp = 0;
                    if (numbers[0] == numbers[1])
                    {
                        temp = 1;
                    }
                    WriteMemory(writeAddress, temp);
                    break;


                case Opcode.AdjustRelativeBase:
                    RelativeBasePointer += numbers[0];
                    break;


                case Opcode.Halt:
                    State = State.Halt;
                    return State;
            }

           
            InstructionPointer += instruction.Size;
            return State;
        }
        #endregion
    }
}
